Code/Yospace/Source/YospaceID3MetadataHandler.swift

```diff
@@ -0,0 +1,95 @@
+//
+//  YospaceID3MetadataHandler.swift
+//
+//  Created by Raffi on 30/05/2024.
+//  Copyright © 2024 THEOplayer. All rights reserved.
+//
+
+import THEOplayerSDK
+import YOAdManagement
+
+class YospaceID3MetadataHandler {
+    private let player: THEOplayerSDK.THEOplayer
+    private let session: YOSession
+    private var addTrackEventListener: THEOplayerSDK.EventListener?
+    private var addCueEventListener: THEOplayerSDK.EventListener?
+    private var metadata: YospaceTimedMetadata = .init()
+
+    init(player: THEOplayerSDK.THEOplayer, session: YOSession) {
+        self.player = player
+        self.session = session
+        self.addTrackEventListener = self.player.textTracks.addEventListener(type: THEOplayerSDK.TextTrackListEventTypes.ADD_TRACK) { [weak self] event in
+            self?.addTrackEventHandler(event: event)
+        }
+    }
+
+    private func addTrackEventHandler(event: THEOplayerSDK.AddTrackEvent) {
+        if let textTrack: THEOplayerSDK.TextTrack = event.track as? THEOplayerSDK.TextTrack,
+           textTrack.kind == THEOplayerSDK.TextTrackKind.metadata._rawValue,
+           textTrack.type == "id3" {
+            self.addCueEventListener = textTrack.addEventListener(type: THEOplayerSDK.TextTrackEventTypes.ADD_CUE) { [weak self] event in
+                self?.addCueEventHandler(event: event)
+            }
+        }
+    }
+
+    private func addCueEventHandler(event: THEOplayerSDK.AddCueEvent) {
+        if let key: String = event.cue.contentDictionary?["id"],
+           let value: String = event.cue.contentDictionary?["text"] {
+            let _value: String = String(value[value.index(after: value.startIndex)..<value.endIndex])
+            if key == "YMID" {
+                if !self.metadata.ymid.isEmpty {
+                    self.metadata = YospaceTimedMetadata()
+                }
+                self.metadata.ymid = _value
+            } else if key == "YTYP" {
+                self.metadata.ytyp = _value
+            } else if key == "YSEQ" {
+                self.metadata.yseq = _value
+            } else if key == "YDUR" {
+                self.metadata.ydur = _value
+            }
+        }
+
+        if let metadata: YOTimedMetadata = YOTimedMetadata.create(withMediaId: self.metadata.ymid, sequence: self.metadata.yseq, type: self.metadata.ytyp, offset: self.metadata.ydur, playhead: self.player.currentTime),
```

Comment:
Use `event.cue.startTime` instead of `player.currentTime`. (But again, only on `entercue` or `cuechange`. 😉)

Reply:
Done